File: scripts/reference-solutions/contiguous-array.py
import sys


def main():
    data = sys.stdin.read().split()
    n = int(data[0])
    nums = list(map(int, data[1 : 1 + n]))
    first = {0: -1}
    prefix = 0
    best = 0
    for i, x in enumerate(nums):
        prefix += 1 if x == 1 else -1
        if prefix in first:
            best = max(best, i - first[prefix])
        else:
            first[prefix] = i
    print(best)


main()

File: scripts/reference-solutions/contiguous-array.cpp
#include <algorithm>
#include <array>
#include <climits>
#include <deque>
#include <functional>
#include <iostream>
#include <set>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int n;
  cin >> n;
  unordered_map<int, int> first;
  first.reserve(n * 2 + 1);
  first[0] = -1;

  int prefix = 0;
  int best = 0;
  for (int i = 0; i < n; ++i) {
    int value;
    cin >> value;
    prefix += value == 1 ? 1 : -1;
    auto it = first.find(prefix);
    if (it == first.end()) {
      first[prefix] = i;
    } else {
      best = max(best, i - it->second);
    }
  }
  cout << best << '\n';
  return 0;
}
